### PR TITLE
ABC boundary

### PR DESCRIPTION
**Pull request by dbochkov-flexcompute**
- Introduces the ability to set first order absorbing boundary conditions on simulation domain boundaries using either `ABCBoundary` or `ModeABCBoundary`. 
  - In case of `ABCBoundary` the user my provide effective permittivity and conductivity to use in one-way wave equation. Otherwise, Tidy3D will autodetect the meidum on the boundary and use its parameters (if multiple mediums or anisotropic medium, it will throw an error)
  - In case of `ModeABCBoundary` the user need to specify the plane and mode spec for finding the mode to absorb. If there is already a modal source or a monitor, `ModeABCBoundary.from_source()` and `ModeABCBoundary.from_monitor()` can be used.
- Introduces field `frame` in `ModeSource` (default=None) and `WavePort` (default: `PECFrame()`), which could currently be set to `PECFrame(length=...)` to automatically place a thin PEC frame of a given length around them.
  - Automatic creation of frames is done in `Simulation.with_port_frames`, which could be used to visualize where they are placed exactly. This is what backend also uses for running actual simulation
- Introduces `PortAbsorber` class that represent a plane inside the simulation domain where first order absorbing boundary conditions will be placed.
  - Port absorbers are passed to `Simulation` through the field `absorbers`.
  - In simulation plots they are drawn similar to sources but in blue color.
  - `WavePort` now have a field `absorber: bool = True` that allows to automatically put an absorber in the port.
  - Each `PortAbsorber` automatically gets a pec frame of length one and also PEC backing to separate absorbed field from surroundings.
- Adds an option in `ModeMonitors` and `WavePort` to use non-conjugated dot product for mode decomposition with defaults 

backend https://github.com/flexcompute/compute/pull/1445 

Screenshot_20250719_114118

## Greptile Summary

This PR introduces comprehensive support for Absorbing Boundary Conditions (ABC) in Tidy3D electromagnetic simulations. The implementation adds first-order absorbing boundary conditions as an alternative to PML boundaries, particularly useful for scenarios involving dispersive materials or structures intersecting boundaries at angles where PML might cause divergence issues.

The key additions include three main boundary condition types:
- `ABCBoundary`: Standard ABC with optional permittivity/conductivity parameters, with automatic medium detection when parameters aren't provided
- `ModeABCBoundary`: Mode-specific ABC that absorbs specific waveguide modes based on plane and mode specifications
- `InternalAbsorber`: Absorbing planes placed inside the simulation domain rather than at boundaries

The PR also introduces automatic PEC frame functionality through the `PECFrame` class, which can be placed around `ModeSource` and `WavePort` objects to improve injection quality. Wave ports now support automatic absorber placement via the `absorber` boolean field, and both `ModeMonitor` and `WavePort` gain a `conjugated_dot_product` field for controlling mode decomposition calculations.

The implementation integrates seamlessly with existing simulation infrastructure, providing factory methods for creating ABC boundaries from existing sources and monitors, comprehensive validation for medium compatibility (requiring homogeneous, isotropic materials), and consistent visualization with dedicated blue coloring for absorber components. The changes extend across multiple simulation types including standard FDTD, EME, and mode simulations, with proper support in the S-matrix component modeler for automatic port absorber placement.

**PR Description Notes:**
- Line 2: Typo "meidum" should be "medium"
- Line 2: Typo "my" should be "may"

## Important Files Changed

<details>
<summary>Files Modified</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| `tidy3d/components/boundary.py` | 4/5 | Introduces comprehensive ABC boundary implementation with three main classes and factory methods |
| `tidy3d/components/simulation.py` | 4/5 | Adds internal absorbers field, ABC validation, plotting methods, and automatic PEC frame generation |
| `tests/test_components/test_absorbers.py` | 4/5 | Comprehensive test suite for all ABC functionality including edge cases and validation |
| `tidy3d/plugins/smatrix/ports/wave.py` | 4/5 | Adds frame support, absorber functionality, and fixes flux calculations using complex values |
| `tidy3d/plugins/smatrix/component_modelers/terminal.py` | 4/5 | Implements automatic port absorber creation and positioning in S-matrix simulations |
| `tidy3d/plugins/smatrix/component_modelers/base.py` | 4/5 | Refactors port shifting logic to use centralized simulation method |
| `tidy3d/components/source/field.py` | 4/5 | Adds optional frame field to ModeSource for automatic PEC frame placement |
| `tidy3d/components/monitor.py` | 4/5 | Adds conjugated_dot_product field for mode decomposition control (has typo in description) |
| `tests/test_components/test_source_frames.py` | 4/5 | Tests frame functionality and collision detection with projection monitors |
| `tidy3d/components/source/frame.py` | 4/5 | Defines abstract and concrete PEC frame classes for source frames |
| `tidy3d/components/mode/simulation.py` | 2/5 | Incomplete implementation with empty tuple type for internal_absorbers field |
| `tidy3d/components/eme/simulation.py` | 4/5 | Adds internal_absorbers field for EME simulation support |
| `tidy3d/__init__.py` | 5/5 | Properly exposes new ABC and frame classes through public API |
| `tidy3d/components/viz/__init__.py` | 5/5 | Adds visualization support for new absorber components |
| `tidy3d/components/viz/plot_params.py` | 5/5 | Defines plot parameters for absorber and ABC visualization |
| `tidy3d/components/viz/styles.py` | 5/5 | Adds blue arrow color constant for absorber visualization |
| `tests/test_components/test_source.py` | 5/5 | Adds comprehensive tests for PECFrame functionality and validation |

</details>

## Confidence score: 3/5

- This PR introduces significant new functionality but has several implementation issues that need attention before merging.
- The score reflects concerns about incomplete implementations (particularly in mode simulation), potential type inconsistencies, and some validation gaps that could cause runtime issues.
- Files requiring attention: `tidy3d/components/mode/simulation.py` (incomplete absorber implementation), `tidy3d/components/monitor.py` (typo in field description), and the overall integration testing of ABC boundaries with various medium types.

## Sequence Diagram
```mermaid
sequenceDiagram
    participant User
    participant Simulation
    participant ABCBoundary
    participant ModeABCBoundary
    participant InternalAbsorber
    participant ModeSolver
    participant Grid
    participant Validator

    User->>ABCBoundary: Create ABCBoundary(permittivity=2.0)
    ABCBoundary->>ABCBoundary: Validate permittivity > 0
    
    User->>ModeABCBoundary: Create ModeABCBoundary(plane=Box, mode_spec=ModeSpec)
    ModeABCBoundary->>Validator: Validate plane is 2D
    ModeABCBoundary->>ModeSpec: Validate mode_spec parameters
    
    User->>InternalAbsorber: Create InternalAbsorber(boundary_spec=ABCBoundary)
    InternalAbsorber->>Validator: Validate boundary_spec has permittivity
    InternalAbsorber->>Validator: Validate size is planar
    
    User->>Simulation: Create Simulation with ABC boundaries
    Simulation->>Validator: Validate ABC boundaries are homogeneous
    Simulation->>Validator: Check medium compatibility with ABC
    Simulation->>Validator: Validate internal absorbers placement
    
    Simulation->>Grid: Generate grid with absorber considerations
    Grid->>InternalAbsorber: Calculate shifted positions based on grid_shift
    
    Simulation->>ModeSolver: Setup mode solving for ModeABCBoundary
    ModeSolver->>ModeABCBoundary: Extract mode parameters
    ModeSolver->>ModeSolver: Solve for absorbing modes
    
    Simulation->>Simulation: Finalize simulation with frames and absorbers
    Simulation->>User: Return validated simulation ready for execution
```

### CODE REVIEW

**Comment by weiliangjin2021** [comment]
could you drop the "revert ..." commit to have a cleaner PR?

**Comment by dbochkov-flexcompute** [comment]
> could you drop the "revert ..." commit to have a cleaner PR?

ah, yeah, forgot about that

File: tidy3d/components/boundary.py [hunk 1]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
     """Perfect magnetic conductor boundary condition class."""
 
 
+# ABC keyword
+class AbstractABCBoundary(BoundaryEdge, ABC):
+    """One-way wave equation absorbing boundary conditions abstract base class."""
+
+    small_conductivity_approx: bool = pd.Field(
```
Comment (weiliangjin2021) [inline tidy3d/components/boundary.py]: to be removed since it makes little difference?

File: tidy3d/components/boundary.py [hunk 2]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
...
+    )
+
+
+class ABCBoundary(AbstractABCBoundary):
+    """One-way wave equation absorbing boundary conditions."""
+
+    permittivity: Optional[pd.PositiveFloat] = pd.Field(
+        None,
+        title="Effective Permittivity",
+        description="Enforced effective permittivity.",
```
Comment (weiliangjin2021) [inline tidy3d/components/boundary.py]: Add to description on the meaning of "None".

File: tidy3d/components/boundary.py [hunk 3]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
...
+        "frequency.",
+    )
+
+
+class ABCBoundary(AbstractABCBoundary):
+    """One-way wave equation absorbing boundary conditions."""
+
+    permittivity: Optional[pd.PositiveFloat] = pd.Field(
+        None,
+        title="Effective Permittivity",
```
Comment (weiliangjin2021) [inline tidy3d/components/boundary.py:100]: maybe "Effective permittivity for determining propagation constant"?

File: tidy3d/components/boundary.py [hunk 4]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
...
+        title="Effective Conductivity",
+        description="Enforced effective conductivity.",
+    )
+
+    @pd.validator("conductivity", always=True)
+    @skip_if_fields_missing(["permittivity"])
+    def _conductivity_only_with_float_permittivity(cls, val, values):
+        """Validate that conductivity can be provided only with float permittivity."""
+        perm = values["permittivity"]
+        if val is not None and not isinstance(perm, float):
```
Comment (weiliangjin2021) [inline tidy3d/components/boundary.py]: `if perm is None and val is not None`

File: tidy3d/components/boundary.py [hunk 5]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
...
+                "Field 'conductivity' in 'ABCBoundary' can only be provided "
+                "simultaneously with 'permittivity'."
+            )
+        return val
+
+
+class ModeABCBoundary(AbstractABCBoundary):
+    """One-way wave equation absorbing boundary conditions for absorbing a waveguide mode."""
+
+    small_conductivity_approx: bool = pd.Field(
```
Comment (weiliangjin2021) [inline tidy3d/components/boundary.py]: to be removed as well

File: tidy3d/components/simulation.py [hunk 6]
```diff
@@ -3127,6 +3155,94 @@ def _projection_monitors_homogeneous(cls, val, values):
...
+            structures=values.get("structures") or [],
+        )
+
+        with log as consolidated_logger:
+            for mediums in mediums_all_sides:
+                if mediums is not None:
+                    # make sure there is no more than one medium in the returned list
+                    if len(mediums) > 1:
+                        raise SetupError(
+                            f"{len(mediums)} different mediums detected on an 'ABCBoundary'. Boundary must be homogeneous."
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py:3514]: Add to error message (and the others below) that the other option is to provide a float permittivity to ABC

**Review by weiliangjin2021 — APPROVED** [review]

File: tidy3d/components/boundary.py [hunk 7]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
...
+        title="Small Conductivity Approximation",
+        description="If ``False`` then the effective permettivity ``eps`` in the one-wave equation"
+        " is modified such that the equation exactly satisfy wave propagation at the central"
+        " frequency.",
+    )
+
+    mode_spec: ModeSpec = pd.Field(
+        ModeSpec(),
+        title="Mode Specification",
+        description="Parameters to feed to mode solver which determine modes.",
```
Comment (dmarek-flex) [inline tidy3d/components/boundary.py]: > Parameters that determine the modes computed by the mode solver.

slight improvement

File: tidy3d/components/boundary.py [hunk 8]
```diff
@@ -48,6 +51,174 @@ class PMCBoundary(BoundaryEdge):
...
+    )
+
+    mode_index: pd.NonNegativeInt = pd.Field(
+        0,
+        title="Mode Index",
+        description="Index into the collection of modes returned by mode solver. "
+        " Specifies which mode to absorbed using these boundary conditions. "
```
Comment (dmarek-flex) [inline tidy3d/components/boundary.py]: whitespace typo, and slightly improved

> The absorbing boundary conditions are configured to absorb the specified mode.

File: tidy3d/components/boundary.py (hunk 7 above)
Comment (dmarek-flex) [reply tidy3d/components/boundary.py]: and also modify below in the helper functions

**Review by dmarek-flex — APPROVED** [review]
Looks good to me! A couple suggestions to improve descriptions. Also is there a simple reference you could add that describes the BC? like Taflove?

**Comment by dbochkov-flexcompute** [comment]
This PR needs a re-review, please see the update first message

@momchil-flex fyi regarding non-conjugate mode decomposition option

File: tidy3d/components/simulation.py [hunk 9]
```diff
@@ -357,6 +350,28 @@ class AbstractYeeGridSimulation(AbstractSimulation, ABC):
...
+    )
+
+    post_norm: Union[float, FreqDataArray] = pydantic.Field(
+        1.0,
+        title="Post Normalization Values",
+        description="Factor to multiply the fields by after running, "
+        "given the adjoint source pipeline used. Note: this is used internally only.",
+    )
+
+    absorbers: tuple[PortAbsorber, ...] = pydantic.Field(
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py]: how about calling it `internal_absorbers", to distinguish them from boundary absorbers.

File: tidy3d/components/simulation.py [hunk 10]
```diff
@@ -357,6 +350,28 @@ class AbstractYeeGridSimulation(AbstractSimulation, ABC):
...
+        (),
+        title="Port Absorbers",
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py]: no need to mention about "port" in FDTD solver level?

File: tidy3d/components/simulation.py [hunk 11]
```diff
@@ -1973,6 +2081,65 @@ def _invalidate_solver_cache(self) -> None:
...
+        if name is None:
+            name = f"A '{obj.type}'"
+
+        # get the grid boundaries and sizes along port normal from the simulation
+        normal_axis = obj._normal_axis
+        grid_boundaries = self.grid.boundaries.to_list[normal_axis]
+        grid_centers = self.grid.centers.to_list[normal_axis]
+
+        # get the index of the grid cell where the port lies
+        port_position = obj.center[normal_axis]
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py]: again, shouldn't be named "port" as it's a generic method

File: tidy3d/components/simulation.py [hunk 12]
```diff
@@ -1973,6 +2081,65 @@ def _invalidate_solver_cache(self) -> None:
...
+        port_pos_gt_grid_bounds = np.argwhere(port_position > grid_boundaries)
+
+        # no port index can be determined
+        if len(port_pos_gt_grid_bounds) == 0:
+            raise SetupError(f"Port position '{port_position}' outside of simulation bounds.")
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py]: it only errors if port_pos is to the left of the simulation domain, but not the right. Should it either error in both cases, or shift to the simulation domain boundary in both cases? The former sounds better?

**Review by weiliangjin2021 — COMMENTED** [review]
Very exciting bundle of new features! Half way through.

Regarding `Simulation.with_port_frames` to be picked up as the actual simulation, shall we define a method for a more generic purpose, to include the previous `volumetric_structures` as well?

**Comment by weiliangjin2021** [comment]
@yuanshen-flexcompute is it a push by mistake?

**Comment by yuanshen-flexcompute** [comment]
Yes sorry! I am trying to access this branch to run a notebook.

**Comment by dbochkov-flexcompute** [comment]
> Yes sorry! I am trying to access this branch to run a notebook.

no worries, reverted back

**Comment by weiliangjin2021** [comment]
> each PortAbsorber automatically gets a pec frame of length one 

 automatically gets a pec frame of length two?

File: tidy3d/components/boundary.py [hunk 13]
```diff
@@ -73,6 +86,235 @@ class PMCBoundary(BoundaryEdge):
...
+class PortAbsorber(Box):
+    """Internally placed plane with one-way wave equation boundary conditions for absorption of electromagnetic waves."""
+
+    direction: Direction = pd.Field(
+        ...,
+        title="Absorption Direction",
+        description="Direction in which field is absorbed.",
+    )
+
+    shift: int = pd.Field(
```
Comment (weiliangjin2021) [inline tidy3d/components/boundary.py]: `grid_shift` for clarity?

File: tidy3d/components/source/frame.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+from tidy3d.components.base import Tidy3dBaseModel
+
+
+class AbstractSourceFrame(Tidy3dBaseModel, ABC):
+    """Abstract base class for all source frames."""
+
+    length: int = pydantic.Field(
+        2,
+        title="Length",
+        description="Length of the frame in number of cells.",
```
Comment (weiliangjin2021) [inline tidy3d/components/source/frame.py]: more docstring explaining the direction of length?

File: tidy3d/components/source/field.py [hunk 15]
```diff
@@ -403,6 +404,12 @@ class ModeSource(DirectionalSource, PlanarSource, BroadbandSource):
         "``num_modes`` in the solver will be set to ``mode_index + 1``.",
     )
 
+    frame: Optional[PECFrame] = pydantic.Field(
+        None,
+        title="Source Frame.",
+        description="Add a thin frame around the source during FDTD run for an improved injection.",
```
Comment (weiliangjin2021) [inline tidy3d/components/source/field.py]: elaborate on where exactly is the frame?

File: tidy3d/components/simulation.py [hunk 16]
```diff
@@ -5318,3 +5664,86 @@ def from_scene(cls, scene: Scene, **kwargs) -> Simulation:
...
 
     _boundaries_for_zero_dims = validate_boundaries_for_zero_dims()
+
+    def _make_pec_frame(self, obj) -> Structure:
+        """Make a pec frame around a mode source."""
+        span_inds = np.array(self.grid.discretize_inds(obj))
+
+        coords = self.grid.boundaries.to_list
+        direction = obj.direction
+        if isinstance(obj, ModeSource):
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py:5645]: Right now, `obj` can be either a mod source, or port absorber?

**Comment by weiliangjin2021** [comment]
> > each PortAbsorber automatically gets a pec frame of length one
> 
> automatically gets a pec frame of length two?

Ah i see where it is implemented. So why is the default length 2 for mode solver, and 1 for port absorber?

**Review by weiliangjin2021 — COMMENTED** [review]
A few more minor comments

File: tidy3d/components/simulation.py (hunk 9 above)
Comment (dbochkov-flexcompute) [reply tidy3d/components/simulation.py]: yeah, I was going back and forth between `internal_absorbers`/`InternalAbsorber`, `absorbers`/`Absorber`, and `port_absorbers`/`PortAbsorber` (using port as a loose synonym to internal). I think I will convert everything into `internal_absorbers`/`InternalAbsorber`

File: tidy3d/components/simulation.py [hunk 17]
```diff
@@ -3239,6 +3457,122 @@ def _projection_monitors_homogeneous(cls, val, values):
 
         return val
 
+    @classmethod
+    def _get_mediums_on_abc(
+        cls, boundary_spec, medium, center, size, structures
```
Comment (dmarek-flex) [inline tidy3d/components/simulation.py]: just pass `Box` instead of `center` and `size`?

File: tidy3d/components/simulation.py [hunk 18]
```diff
@@ -3239,6 +3486,96 @@ def _projection_monitors_homogeneous(cls, val, values):
...
+
+        with log as consolidated_logger:
+            for mediums in mediums_all_sides:
+                if mediums is not None:
+                    # make sure there is no more than one medium in the returned list
+                    if len(mediums) > 1:
+                        raise SetupError(
+                            f"{len(mediums)} different mediums detected on an 'ABCBoundary'. Boundary must be homogeneous."
+                            "Alternatively, effective permeability and conductivity can be directly provided as "
+                            "parameters for an 'ABCBoundary', in which case this medium check is skipped."
```
[the remark this answers is not in the document]
Comment (yaugenst-flex) [reply tidy3d/components/simulation.py:3516]: it does use single quotes

File: tidy3d/components/source/field.py [hunk 19]
```diff
@@ -403,6 +404,14 @@ class ModeSource(DirectionalSource, PlanarSource, BroadbandSource):
         "``num_modes`` in the solver will be set to ``mode_index + 1``.",
     )
 
+    frame: Optional[PECFrame] = pydantic.Field(
+        None,
+        title="Source Frame.",
```
Comment (dmarek-flex) [inline tidy3d/components/source/field.py]: remove period

File: tidy3d/components/boundary.py [hunk 20]
```diff
@@ -73,6 +86,235 @@ class PMCBoundary(BoundaryEdge):
     """Perfect magnetic conductor boundary condition class."""
 
 
+# ABC keyword
```
Comment (dmarek-flex) [inline tidy3d/components/boundary.py]: unneeded comment?

File: tidy3d/components/boundary.py [hunk 21]
```diff
@@ -73,6 +86,235 @@ class PMCBoundary(BoundaryEdge):
...
+
+    permittivity: Optional[pd.PositiveFloat] = pd.Field(
+        None,
+        title="Effective Permittivity",
+        description="Effective permittivity for determining propagation constant. "
+        "If ``None``, this value will be automatically inferred from the medium at "
+        "the domain boundary and the central frequency of the source.",
+    )
+
+    conductivity: Optional[pd.NonNegativeFloat] = pd.Field(
```
Comment (dmarek-flex) [inline tidy3d/components/boundary.py:107]: add units (I think S/um is what we use for `Medium`)

File: tidy3d/components/boundary.py [hunk 22]
```diff
@@ -73,6 +86,235 @@ class PMCBoundary(BoundaryEdge):
...
+        )
+
+
+class InternalAbsorber(Box):
+    """Internally placed plane with one-way wave equation boundary conditions for absorption of electromagnetic waves."""
+
+    direction: Direction = pd.Field(
+        ...,
+        title="Absorption Direction",
+        description="Direction in which field is absorbed.",
```
Comment (dmarek-flex) [inline tidy3d/components/boundary.py]: Maybe a bit clearer to mention waves:
"Indicates which direction of traveling waves are absorbed."

File: tidy3d/components/boundary.py [hunk 23]
```diff
@@ -73,6 +86,235 @@ class PMCBoundary(BoundaryEdge):
...
+class AbstractABCBoundary(BoundaryEdge, ABC):
+    """One-way wave equation absorbing boundary conditions abstract base class."""
+
+
+class ABCBoundary(AbstractABCBoundary):
+    """One-way wave equation absorbing boundary conditions.
+    See, for example, John B. Schneider, Understanding the Finite-Difference Time-Domain Method, Chapter 6.
+    """
+
+    permittivity: Optional[pd.PositiveFloat] = pd.Field(
```
Comment (dmarek-flex) [inline tidy3d/components/boundary.py]: Can `permittivity` be less than 1?

File: tidy3d/components/simulation.py [hunk 24]
```diff
@@ -1750,6 +1868,9 @@ def subsection(
             Used internally.
         deep_copy: bool = True
             Recursively copy all nested objects in the generated simulation object.
+        internal_absorbers : Tuple[SourceType, ...] = None
```
Comment (dmarek-flex) [inline tidy3d/components/simulation.py]: wrong type hint `SourceType`

File: tidy3d/components/simulation.py [hunk 25]
```diff
@@ -1973,6 +2100,73 @@ def _invalidate_solver_cache(self) -> None:
...
+                    f"{name} normal is less than 2 cells to the boundary "
+                    f"on +{'xyz'[normal_axis]} side."
+                    "Please either increase the mesh resolution near the obj or "
+                    "move the obj away from the boundary."
+                )
+
+        new_pos = grid_centers[shifted_index]
+        return new_pos - obj_position
+
+    def _shift_object(self, obj: InternalAbsorber, direction: Direction, shift: int) -> float:
```
Comment (dmarek-flex) [inline tidy3d/components/simulation.py]: Maybe better as a method for `InternalAbsorber`? If it is truly generalizable to other `Geometry`, probably could go in the components geometry part of the code base.

Comment (dmarek-flex) [reply tidy3d/components/simulation.py]: also the return type hint is incorrect

File: tidy3d/components/simulation.py (hunk 16 above)
Comment (dbochkov-flexcompute) [reply tidy3d/components/simulation.py:5645]: yes, added a type annotation for that

File: tidy3d/components/simulation.py [hunk 26]
```diff
@@ -1981,6 +2108,73 @@ def _invalidate_solver_cache(self) -> None:
         """Clear cached attributes that become stale when subpixel changes."""
         self._cached_properties.pop("_mode_solver", None)
 
+    def _shift_value_signed(
```
Comment (dmarek-flex) [inline tidy3d/components/simulation.py]: I think this could live in Geometry part of the code base. In `geometry/utils.py` could be a good place. I think you only need the grid from the `Simulation`.

**Review by dmarek-flex — APPROVED** [review]
Thanks, LGTM! Still trying to find to time to test it out myself, but the PR looks good.

File: tidy3d/components/boundary.py (hunk 23 above)
Comment (dbochkov-flexcompute) [reply tidy3d/components/boundary.py]: added a restriction

File: tidy3d/components/simulation.py (hunk 25 above)
Comment (dbochkov-flexcompute) [reply tidy3d/components/simulation.py]: moved to utils

File: tidy3d/components/simulation.py (hunk 26 above)
Comment (dbochkov-flexcompute) [reply tidy3d/components/simulation.py]: moved to utils

**Review by weiliangjin2021 — APPROVED** [review]

File: tidy3d/components/eme/simulation.py [hunk 27]
```diff
@@ -209,6 +209,12 @@ class EMESimulation(AbstractYeeGridSimulation):
         "use 'smatrix_in_basis' to use another set of modes or input field.",
     )
 
+    internal_absorbers: tuple[()] = pd.Field(
```
[the remark this answers is not in the document]
Comment (yaugenst-flex) [reply tidy3d/components/eme/simulation.py:212]: @dbochkov-flexcompute can you confirm that the `tuple[()]` type hint is intentional? The field currently only accepts empty tuples? This is not how I see it being used in tests...

[the remark this answers is not in the document]
Comment (yaugenst-flex) [reply tidy3d/components/eme/simulation.py:212]: Aaah sorry nevermind, didn't see that this was in EME... in Pydantic v2 there will be nicer ways to disable fields

[the remark this answers is not in the document]
Comment (dbochkov-flexcompute) [reply tidy3d/components/eme/simulation.py:212]: @yaugenst-flex yeah, I was just following the same treatment as for sources, but maybe I got a bit too lazy there. The primary reason for having internal_absorbers defined at AbstractYeeCellSimulation was an easier way for changing plotting functions. Maybe another option is to have a generic plot_extras() at AbstractYeeCellSimulation level, which can include any specific to Simulation, EMESimulation, and ModeSimulation elements. However, in that way I don't see a good way to specify plotting parameters for those additional elements in plot()

File: tidy3d/components/simulation.py [hunk 28]
```diff
@@ -5326,3 +5624,91 @@ def from_scene(cls, scene: Scene, **kwargs) -> Simulation:
         )
 
     _boundaries_for_zero_dims = validate_boundaries_for_zero_dims()
+
+    def _make_pec_frame(self, obj: Union[ModeSource, InternalAbsorber]) -> Structure:
+        """Make a pec frame around a mode source."""
```
Comment (weiliangjin2021) [inline tidy3d/components/simulation.py]: would be good to add in the docstring (also docstring of Simulation absorber fields, and absorber docstring) that both PEC frame and backing are added.
